aws/resource_aws_cloudhsm2_hsm.go

```diff
@@ -0,0 +1,285 @@
+			},
+		},
+	}
+}
+
+func resourceAwsCloudHsm2HsmImport(
+	d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	d.Set("hsm_id", d.Id())
+	return []*schema.ResourceData{d}, nil
+}
+
+func describeHsm(d *schema.ResourceData, meta interface{}) (*cloudhsmv2.Hsm, error) {
+	conn := meta.(*AWSClient).cloudhsmv2conn
+
+	clusterId := d.Get("cluster_id").(string)
+	hsmId := d.Id()
+	filters := []*string{&clusterId}
+	result := int64(1)
+	out, err := conn.DescribeClusters(&cloudhsmv2.DescribeClustersInput{
+		Filters: map[string][]*string{
+			"clusterIds": filters,
+		},
+		MaxResults: &result,
+	})
+	if err != nil {
+		log.Printf("[WARN] Error on retrieving CloudHSMv2 Cluster (%s) when waiting: %s", d.Id(), err)
+		return nil, err
+	}
+
+	var cluster *cloudhsmv2.Cluster
+
+	for _, c := range out.Clusters {
+		if *c.ClusterId == clusterId {
+			cluster = c
+		}
+	}
+
+	var hsm *cloudhsmv2.Hsm
+	if cluster != nil {
+		for _, h := range cluster.Hsms {
+			if *h.HsmId == hsmId {
+				hsm = h
+			}
+		}
+	}
+
+	return hsm, nil
+}
+
+func resourceAwsCloudHsm2HsmRefreshFunc(
+	d *schema.ResourceData, meta interface{}) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		hsm, err := describeHsm(d, meta)
+
+		if hsm == nil {
+			return 42, "destroyed", nil
+		}
+
+		if hsm.State != nil {
+			log.Printf("[DEBUG] CloudHSMv2 Cluster status (%s): %s", d.Id(), *hsm.State)
+		}
+
+		return hsm, *hsm.State, err
+	}
+}
+
+func resourceAwsCloudHsm2HsmCreate(d *schema.ResourceData, meta interface{}) error {
+	cloudhsm2 := meta.(*AWSClient).cloudhsmv2conn
+
+	clusterId := d.Get("cluster_id").(string)
+
+	cluster, err := describeCloudHsm2Cluster(clusterId, meta)
+
+	if cluster == nil {
+		log.Printf("[WARN] Error on retrieving CloudHSMv2 Cluster: %s %s", clusterId, err)
+		return err
+	}
+
+	availabilityZone := d.Get("availability_zone").(string)
+	if len(availabilityZone) == 0 {
+		subnetId := d.Get("subnet_id").(string)
+		for az, sn := range cluster.SubnetMapping {
+			if *sn == subnetId {
+				availabilityZone = az
+			}
+		}
+	}
+
+	input := &cloudhsmv2.CreateHsmInput{
+		ClusterId:        aws.String(clusterId),
+		AvailabilityZone: aws.String(availabilityZone),
+	}
+
+	ipAddress := d.Get("ip_address").(string)
+	if len(ipAddress) != 0 {
+		input.IpAddress = aws.String(ipAddress)
+	}
+
+	log.Printf("[DEBUG] CloudHSMv2 HSM create %s", input)
+
+	var output *cloudhsmv2.CreateHsmOutput
+
+	errRetry := resource.Retry(180*time.Second, func() *resource.RetryError {
+		var err error
+		output, err = cloudhsm2.CreateHsm(input)
+		if err != nil {
+			if isAWSErr(err, cloudhsmv2.ErrCodeCloudHsmInternalFailureException, "request was rejected because of an AWS CloudHSM internal failure") {
+				log.Printf("[DEBUG] CloudHSMv2 HSM re-try creating %s", input)
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if errRetry != nil {
+		return errRetry
```

Comment:
We should provide context about this error here, e.g. `return fmt.Errorf("error creating CloudHSMv2 HSM: %s", errRetry)`